crates/shared/src/trade_finding.rs

```diff
@@ -49,6 +120,110 @@ pub struct Trade {
     pub tx_origin: Option<H160>,
 }
 
+/// A trade with JIT orders.
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct Trade {
+    pub clearing_prices: HashMap<H160, U256>,
+    /// How many units of gas this trade will roughly cost.
+    pub gas_estimate: Option<u64>,
+    pub pre_interactions: Vec<Interaction>,
+    /// Interactions needed to produce the expected trade amount.
+    pub interactions: Vec<Interaction>,
+    /// Which solver provided this trade.
+    pub solver: H160,
+    /// If this is set the quote verification need to use this as the
+    /// `tx.origin` to make the quote pass the simulation.
+    pub tx_origin: Option<H160>,
+    pub jit_orders: Vec<dto::JitOrder>,
+}
+
+impl Trade {
+    pub fn out_amount(
+        &self,
+        buy_token: &H160,
+        sell_token: &H160,
+        in_amount: &U256,
+        order_kind: &OrderKind,
+    ) -> Result<U256> {
+        let sell_price = self
+            .clearing_prices
+            .get(sell_token)
+            .context("clearing sell price missing")?
+            .to_big_rational();
+        let buy_price = self
+            .clearing_prices
+            .get(buy_token)
+            .context("clearing buy price missing")?
+            .to_big_rational();
+        let order_amount = in_amount.to_big_rational();
+
+        let out_amount = match order_kind {
+            OrderKind::Sell => order_amount
+                .mul(&sell_price)
+                .checked_div(&buy_price)
+                .context("div by zero: buy price")?,
+            OrderKind::Buy => order_amount
+                .mul(&buy_price)
+                .checked_div(&sell_price)
+                .context("div by zero: sell price")?,
+        };
+
+        big_rational_to_u256(&out_amount).context("out amount is not a valid U256")
+    }
+
+    pub fn jit_orders_executed_amounts(&self) -> Result<HashMap<H160, U256>> {
+        let mut executed_amounts: HashMap<H160, U256> = HashMap::new();
+
+        for jit_order in self.jit_orders.iter() {
+            let sell_price = self
+                .clearing_prices
+                .get(&jit_order.sell_token)
+                .context("JIT order sell token clearing price is missing")?
+                .to_big_rational();
+            let buy_price = self
+                .clearing_prices
+                .get(&jit_order.buy_token)
+                .context("JIT order buy token clearing price is missing")?
+                .to_big_rational();
+            let executed_amount = jit_order.executed_amount.to_big_rational();
+            let (executed_sell, executed_buy) = match jit_order.side {
+                Side::Sell => {
+                    let buy_amount = executed_amount
+                        .clone()
+                        .mul(&sell_price)
+                        .checked_div(&buy_price)
```

Comment:
When computing the buy amount on sell orders we need to use `ceil_div()` to round in the trader's favor.